### PR TITLE
Agents.md for AI Agents

### PR DESCRIPTION
**Pull request by memodi**
## Description

Agents.md could be useful to provide context about network-observability operator repo to build, test, document new features or answer questions about existing features.

## Dependencies

n/a

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [ ] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [ ] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labeled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [ ] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+Best practices for AI coding agents on NetObserv Operator.
+
+> **Note**: Symlinked as [CLAUDE.md](CLAUDE.md) for Claude Code auto-loading.
+
+## Project Context
+
+**NetObserv Operator** - Kubernetes/OpenShift operator for network observability
+(operator-sdk)
+
+**Components:**
```
Comment (jotak) [inline AGENTS.md:12]: would it improve the context if we add links to the github repos?

File: AGENTS.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+## Common Task Templates
+
+### Add FlowCollector Field
+```
+Add spec.agent.ebpf.newFeature (bool, default: false):
+1. Update api/flowcollector/v1beta2/flowcollector_types.go (+kubebuilder markers)
+2. Modify internal/controller/ to use field
+3. Add unit tests
+4. Update docs/FlowCollector.md
```
Comment (jotak) [inline AGENTS.md]: doc is automatically updated via make update-bundle (based on doc in code)

File: AGENTS.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+```
+Review for:
+1. Code style consistency
+2. Error handling (wrap with context)
+3. Unit test coverage (Ginkgo/Gomega)
+4. CRD validation markers
+5. Bundle update needed?
+6. Documentation updates
+7. Backward compatibility
+8. Security (RBAC, TLS, input validation)
```
Comment (jotak) [inline AGENTS.md]: 9. Performance, especially memory usage ?

File: AGENTS.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+## Code Review Checklist
+
+```
+Review for:
+1. Code style consistency
+2. Error handling (wrap with context)
+3. Unit test coverage (Ginkgo/Gomega)
+4. CRD validation markers
+5. Bundle update needed?
```
Comment (jotak) [inline AGENTS.md]: IMO this is not important as already covered in checks

File: AGENTS.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  [internal/controller/flowcollector_controller.go](internal/controller/flowcollector_controller.go)
+- FLP:
+  [internal/controller/flp/flp_transfo_reconciler.go](internal/controller/flp/flp_transfo_reconciler.go)
+- Console Plugin Static Config:
+  [internal/controller/consoleplugin/config/static-frontend-config.yaml](internal/controller/consoleplugin/config/static-frontend-config.yaml)
+- Docs: [docs/FlowCollector.md](docs/FlowCollector.md)
+- Sample:
+  [config/samples/flows_v1beta2_flowcollector.yaml](config/samples/flows_v1beta2_flowcollector.yaml)
+
+**Version Info (v1.10.x):**
```
Comment (jotak) [inline AGENTS.md]: Can we avoid writing a version here? (actually, what's the goal? does that mean we should keep it up to date regularly?)

File: AGENTS.md [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+4. Review: "Review for edge cases and errors"
+5. Document: "Update FlowCollector.md"
+6. Test: "Provide test scenarios"
+```
+
+## Contribution Checklist
+
+Before commit:
+1. AI code review
+2. `make build test`
```
Comment (jotak) [inline AGENTS.md]:
```suggestion
2. `make build lint test`
```

File: AGENTS.md (hunk 2 above)
Comment (memodi) [reply AGENTS.md]: is all current documentation in FlowCollector.md generated based on doc in code?

File: AGENTS.md (hunk 1 above)
Comment (memodi) [reply AGENTS.md:12]: Sure, here it could improve those context to provide overall architecture understanding, however in general w.r.t Agents.md the best practice is to use fewer words for agents  to not get confused with lot of context.

File: AGENTS.md (hunk 3 above)
Comment (memodi) [reply AGENTS.md]: good point.

File: AGENTS.md (hunk 5 above)
Comment (memodi) [reply AGENTS.md]: While this may not be the specific reason for it to be updated, Agents.md are expected to be kept updated for newer enhancements and updated context.

I updated this section to remove specific version.

File: AGENTS.md [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+**NetObserv Operator** - Kubernetes/OpenShift operator for network observability
+(operator-sdk)
+
+**Components:**
+- **[eBPF Agent](https://github.com/netobserv/netobserv-ebpf-agent)**: Network flow generation from packets (DaemonSet)
+- **[flowlogs-pipeline](https://github.com/netobserv/flowlogs-pipeline)**: Flow collection, enrichment, export (Deployment/StatefulSet)
+- **[Console Plugin](https://github.com/netobserv/network-observability-console-plugin)**: OpenShift visualization (optional)
+- **CRD**: `FlowCollector` v1beta2 - **single cluster-wide resource named
+  `cluster`**
+- **Integrations**: Loki, Prometheus, Kafka (optional)
```
Comment (stleerh) [inline AGENTS.md]: Loki is also optional.

File: AGENTS.md (hunk 2 above)
Comment (jotak) [reply AGENTS.md]: yes, we never do manual edits there

**Comment by memodi** [comment]
> The challenge is keeping this document up-to-date. I know it's like a README.md for AI agents, but it's quite detailed, so it's easy for this to get outdated. 

@stleerh 
indeed, it's one additional file to maintain besides README.md, however it's not as cumbersome to keep it updated, since you can ask Claude to update it and we can simply verify newer changes. The recommended number of lines for Agents.md is < 400 to provide sufficient context and optimize the token usage. 

> The more it can reference the actual files themselves to get the information, the better.

100%, in this PR it uses actual files wherever possible for e.g.: sections like "Key Files" and "Resources" sections.

**Comment by memodi** [comment]
> Have you tried it ? Do you see improvements ?

@jpinsonneau - I haven't tried specifically to perform code changes in Operator repo but I tried for `Review` since I have below context what to "Review" for, it'd ensure below areas it would evaluate code changes for, whether or not its assessment is actually accurate that's for human behind invoking has to decide 🙃 
```
1. Code style consistency
2. Error handling (wrap with context)
3. Unit test coverage (Ginkgo/Gomega)
4. CRD validation markers
5. Documentation updates
6. Backward compatibility
7. Security (RBAC, TLS, input validation)
8. Performance and Resource utilization, including watching for memory usage impact for large scale clusters.
```

**Comment by memodi** [comment]
are we good to merge this?

**Comment by memodi** [comment]
merged as discussed during scrum.
